include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h

```diff
@@ -66,28 +66,109 @@ gather(const T* input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets
     return sycl::ext::intel::esimd::gather(input + base_offset, offsets * size32<T>, mask);
 }
 
-template <typename T, int N, sycl::access_mode Mode, sycl::access::placeholder P>
+template <typename T, int N, sycl::access_mode Mode, sycl::access::placeholder P, std::enable_if_t<sizeof(T) <= 4, int> = 0>
+sycl::ext::intel::esimd::simd<T, N>
+gather(sycl::accessor<T, 1, Mode, sycl::target::device, P> acc,
+       sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets,
+       ::std::uint32_t base_offset,
+       sycl::ext::intel::esimd::simd_mask<N> mask = 1)
+{
+    // https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory.html#ga32c45e430fa87969f25402fa029a21d3
+    // template <typename T, int N, typename AccessorTy, typename Toffset>
+    // __ESIMD_API
+    // std::enable_if_t<(sizeof(T) <= 4) && (N == 1 || N == 8 || N == 16 || N == 32) && !std::is_pointer<AccessorTy>::value && std::is_integral_v<Toffset>, simd<T, N>>
+    // sycl::_V1::ext::intel::esimd::gather(AccessorTy acc,
+    //                                      simd<Toffset, N> offsets,
+    //                                      uint32_t glob_offset = 0,
+    //                                      simd_mask<N> mask = 1)
+    return sycl::ext::intel::esimd::gather<T>(acc, offsets * size32<T>, base_offset * size32<T>, mask);
+}
+
+template <typename T, int N, sycl::access_mode Mode, sycl::access::placeholder P, std::enable_if_t<sizeof(T) == 2 * sizeof(uint32_t), int> = 0>
 sycl::ext::intel::esimd::simd<T, N>
-gather(sycl::accessor<T, 1, Mode, sycl::target::device, P> input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets,
-       ::std::uint32_t base_offset, sycl::ext::intel::esimd::simd_mask<N> mask = 1)
+gather(sycl::accessor<T, 1, Mode, sycl::target::device, P> acc,
+       sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets,
+       ::std::uint32_t base_offset,
+       sycl::ext::intel::esimd::simd_mask<N> mask = 1)
 {
-    return sycl::ext::intel::esimd::gather<T>(input, offsets * size32<T>, base_offset * size32<T>, mask);
+    // TODO is it correct?
+    using uint32Acc = sycl::accessor<uint32_t, 1, Mode, sycl::target::device, P>;
+
+    uint32Acc*                        __p_uint32_acc = reinterpret_cast<uint32Acc*>(&acc);
```

Comment:
It is very dangerous. It had better not to treat `sycl::accessor` object as a pointer in terms of bit representation; in addition to a pointer `sycl::accessor` encapsulates, it can contain any implementation defined private members.

Also, [ESIMD description](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2023-0/explicit-simd-sycl-extension.html) contains a restrictions section with a note: 
> Unsupported standard SYCL APIs:
	sycl::accessor::get_pointer(). All memory accesses through an accessor are done via explicit APIs.


This is not a memory access, however it is risky. `gather(*__p_uint32_acc,...)` below accesses memory through a pointer converted from an accessor.

Reply:
Sorry, I took a look more closely. It is conversion from one accessor to another (and not to a pointer). Still I am not sure if it is a valid operation.